plugins/woocommerce/src/Internal/Orders/SourceAttributionController.php

```diff
@@ -0,0 +1,473 @@
+use WC_Order;
+use WC_Tracks;
+use WP_User;
+
+/**
+ * Class SourceAttributionController
+ *
+ * @since x.x.x
+ *
+ * phpcs:disable WordPress.PHP.DisallowShortTernary
+ * phpcs:disable Generic.Commenting.DocComment.MissingShort
+ */
+class SourceAttributionController implements RegisterHooksInterface {
+
+	use ScriptDebug;
+	use SourceAttributionMeta {
+		get_prefixed_field as public;
+	}
+
+	/**
+	 * The FeatureController instance.
+	 *
+	 * @var FeaturesController
+	 */
+	private $feature_controller;
+
+	/**
+	 * WooCommerce logger class instance.
+	 *
+	 * @var WC_Logger_Interface
+	 */
+	private $logger;
+
+	/**
+	 * The LegacyProxy instance.
+	 *
+	 * @var LegacyProxy
+	 */
+	private $proxy;
+
+	/**
+	 * Initialization method.
+	 *
+	 * Takes the place of the constructor within WooCommerce Dependency injection.
+	 *
+	 * @internal
+	 *
+	 * @param LegacyProxy         $proxy      The legacy proxy.
+	 * @param FeaturesController  $controller The feature controller.
+	 * @param WC_Logger_Interface $logger     The logger object. If not provided, it will be obtained from the proxy.
+	 */
+	final public function init( LegacyProxy $proxy, FeaturesController $controller, ?WC_Logger_Interface $logger = null ) {
+		$this->proxy              = $proxy;
+		$this->feature_controller = $controller;
+		$this->logger             = $logger ?? $proxy->call_function( 'wc_get_logger' );
+		$this->set_fields_and_prefix();
+	}
+
+	/**
+	 * Register this class instance to the appropriate hooks.
+	 *
+	 * @since x.x.x
+	 * @return void
+	 */
+	public function register() {
+		// Don't run during install.
+		if ( Constants::get_constant( 'WC_INSTALLING' ) ) {
+			return;
+		}
+
+		// Bail if the feature is not enabled.
+		if ( ! $this->feature_controller->feature_is_enabled( 'order_source_attribution' ) ) {
+			return;
+		}
+
+		add_action(
+			'wp_enqueue_scripts',
+			function() {
+				$this->enqueue_scripts_and_styles();
+			}
+		);
+
+		add_action(
+			'admin_enqueue_scripts',
+			function() {
+				$this->enqueue_admin_scripts_and_styles();
+			}
+		);
+
+		// Include our hidden fields on order notes and registration form.
+		$source_form_fields = function() {
+			$this->source_form_fields();
+		};
+
+		add_action( 'woocommerce_after_order_notes', $source_form_fields );
+		add_action( 'woocommerce_register_form', $source_form_fields );
+
+		// Update data based on submitted fields.
+		add_action(
+			'woocommerce_checkout_order_created',
+			function( $order ) {
+				/**
+				 * Run an action to save order source attribution data.
+				 *
+				 * @since x.x.x
+				 *
+				 * @param WC_Order $order The order object.
+				 */
+				do_action( 'woocommerce_order_save_attribution_source_data', $order );
+			}
+		);
+
+		add_action(
+			'woocommerce_order_save_attribution_source_data',
+			function( $order, $data = array() ) {
+				// phpcs:ignore WordPress.Security.NonceVerification
+				$source_data = $this->get_source_values( empty( $data ) ? $_POST : $data );
```

Comment:
Please add a comment on why it's OK to disable that rule here.

Reply:
Added notes about nonce being validated by WooCommerce before reaching this code.
e19137970c270a683a0343d42bdd2f4efd18e30e